django/db/models/sql/datastructures.py

```diff
@@ -79,15 +82,35 @@ def as_sql(self, compiler, connection):
         qn = compiler.quote_name_unless_alias
         qn2 = connection.ops.quote_name
 
+        if hasattr(self.join_field, "get_joining_fields"):
+            join_fields = self.join_field.get_joining_fields()
+        else:
+            warnings.warn(
+                "The usage of get_joining_columns in Join is deprecated. Implement "
+                "get_joining_fields instead.",
+                RemovedInDjango60Warning,
+            )
+            join_fields = self.join_cols
+
         # Add a join condition for each pair of joining columns.
-        for lhs_col, rhs_col in self.join_cols:
+        for lhs, rhs in join_fields:
+            if isinstance(lhs, str):
+                lhs_full_name = "%s.%s" % (qn(self.parent_alias), qn2(lhs))
+                rhs_full_name = "%s.%s" % (qn(self.table_alias), qn2(rhs))
+            else:
+                lhs, rhs = connection.ops.prepare_join_on_clause(
+                    self.parent_alias, lhs, self.table_alias, rhs
+                )
+                lhs_sql, lhs_params = compiler.compile(lhs)
+                lhs_full_name = lhs_sql % lhs_params
+                rhs_sql, rhs_params = compiler.compile(rhs)
+                rhs_full_name = rhs_sql % rhs_params
```

Comment:
I agree that a more conservative posture here would probably be better. Instead of trying to cast types in all cases, let's just cast types in the PostgreSQL case for now 🤔 

I reckon that using `get_joining_fields` is still a nice-to-have, so I'd keep this bit here and rather move some logic to the PostgreSQL `DatabaseOperations`.